File: algorithm-join-point-2/src/main/java/ru/vood/joinpoint2/bpp/JoinPointBeanPostProcessor.java
package ru.vood.joinpoint2.bpp;

import org.springframework.beans.BeansException;
import org.springframework.beans.factory.config.BeanPostProcessor;
import org.springframework.context.ApplicationContext;
import org.springframework.jdbc.core.JdbcTemplate;
import ru.vood.joinpoint2.infrastructure.bean.WorkerBeanInterface;

import java.lang.reflect.Type;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.function.Predicate;

import static java.util.stream.Collectors.toList;

//@Component
public class JoinPointBeanPostProcessor implements BeanPostProcessor {

    private final JdbcTemplate jdbcTemplate;
    private final Map<String, WorkerBeanInterface> beanMap;

    public JoinPointBeanPostProcessor(JdbcTemplate jdbcTemplate, ApplicationContext context, Map<String, WorkerBeanInterface> beanMap) {
        this.jdbcTemplate = jdbcTemplate;
        this.beanMap = beanMap;
    }


    @Override
    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {

        if (beanName.equals("Bean1")) {
            System.out.println(beanName);
            final List<Type> collect = Arrays.stream(bean.getClass().getInterfaces())
                    .filter(new Predicate<Type>() {
                        @Override
                        public boolean test(Type type) {
                            return type == WorkerBeanInterface.class;
                        }
                    })
                    .collect(toList());
            System.out.println(collect);
        }
//        if (beanMap.containsKey(beanName)) {
//            final List<Type> collect = Arrays.stream(bean.getClass().getGenericInterfaces())
//                    .collect(toList());
//            System.out.println(collect);
//        }
        return bean;
    }
}

File: algorithm-join-point/src/main/java/ru/vood/joinpoint/configuration/bpp/JoinPointBeanPostProcessor.kt
package ru.vood.joinpoint.configuration.bpp

import org.springframework.beans.factory.config.BeanPostProcessor
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Component

@Component
class JoinPointBeanPostProcessor(private val jdbcTemplate: JdbcTemplate) : BeanPostProcessor {

    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
/*
        val query = jdbcTemplate.query("select 1 from dual") { rs, rowNum ->
            rs.getString(1)
        }
        println(query)
        if ("Bean1".equals(beanName)) {
            val find = bean.javaClass.genericInterfaces
                    .asSequence()
                    .find { type -> (type as ParameterizedTypeImpl).rawType == BeanInterface::class.java }
            val parameterizedTypeImpl = find as ParameterizedTypeImpl
            parameterizedTypeImpl.actualTypeArguments[1]
            println(find)

            val methods = bean.javaClass.methods
            println(methods)
            val beanExists = jdbcTemplate.query("select DAB.BEAN_ID from DICT_ACT_BEAN DAB where DAB.BEAN_ID=:1",
                    beanName) { rs, rowNum ->
                rs.getString(1)
            }


            if (beanExists.isEmpty())
                throw RuntimeException("Bean $beanName does not exists in table DICT_ACT_BEAN")
            val javaClass = bean.javaClass
            println(bean)
            println(beanName)
        }
*/
        return bean
    }
}